File: HW4/loot.rb
########## IMAN ALI ##########
########## 112204305 #############
########## imaali #############

$pirateColors = [:blue, :green, :purple, :gold]
$maxPlayersAllowed = 5
$minPlayersAllowed = 2
$playerNames = ["Joy", "Nan", "Sat"]

class MerchantShip
  def initialize(value)
    @value = value
  end
  attr_reader :value
end

class PirateShip
  def initialize(color, attack_value)
    @color = color
    @attack_value = attack_value
  end
  attr_reader :color, :attack_value
end

class Captain
  def initialize(color)
    @color = color
  end
  attr_reader :color
end

class Admiral
  @@instance = nil

  def initialize
    if @@instance.nil?
      return self
    else
      raise RuntimeError
    end
  end

  def Admiral.get_instance
    if @@instance.nil?
      @@instance = Admiral.new()
    end
    return @@instance
  end
end

class Player
  def initialize(name)
    @name = name
    @merchant_ships_at_sea = []
    @merchant_ships_captured = []
    @hand = []
    @merchant_pirates = {}
    @dealer = false
  end

  attr_accessor :name, :merchant_ships_at_sea, :merchant_ships_captured, :hand, :merchant_pirates, :dealer


  def deal(game_state)
    @dealer = true
    game_state.deck.cards.shuffle!()
    card = 0
    (1..6).each do
      game_state.players.each do |player|
        current = game_state.deck.cards[card]
        player.hand.append(current)
        game_state.deck.cards.delete(current)
        card += 1
      end
    end
  end

  def draw_card(game)
    if game.deck.cards.length() > 0
      card = game.deck.cards.pop
      @hand.append(card)
    else
        raise RuntimeError
    end
  end

  def float_merchant(card)
    if not card.instance_of?(MerchantShip)
      return false
    end

    if not @hand.include?(card)
      return false
    end

    @hand.delete(card)
    @merchant_ships_at_sea.append(card)
    return true
  end


  def play_pirate(pirate_card, merchant_card, pl)
    if not @hand.include?(pirate_card)
      return false
    end

    if not merchant_card.instance_of?(MerchantShip)
        return false
    end

    if not pl.merchant_ships_at_sea.include?(merchant_card)
      return false
    end

    if not pirate_card.instance_of?(PirateShip)
      return false
    end

    if not pl.merchant_pirates.key?(merchant_card)
      pl.merchant_pirates[merchant_card] = [[self, pirate_card]]
      @hand.delete(pirate_card)
      return true
    end

    last_card = pl.merchant_pirates[merchant_card][- 1][1]
    if last_card.instance_of?(Captain) or last_card.instance_of?(Admiral)
      return false
    end

    pl.merchant_pirates[merchant_card].each do |pair|
      if pair[1].color == pirate_card.color and pair[0] != self
        return false
      end
    end

    pl.merchant_pirates[merchant_card].each do |pair|
      if pair[0] == self and pirate_card.color != pair[1].color
        return false
      end
    end

    pl.merchant_pirates[merchant_card].append([self, pirate_card])
    @hand.delete(pirate_card)
    return true
  end

  def play_captain(captain_card, merchant_card, pl)
    if not @hand.include?(captain_card)
      return false
    end

    if not merchant_card.instance_of?(MerchantShip)
        return false
    end

    if not pl.merchant_ships_at_sea.include?(merchant_card)
      return false
    end

    if not captain_card.instance_of?(Captain)
      return false
    end

    if not pl.merchant_pirates.key?(merchant_card)
      pl.merchant_pirates[merchant_card] = [[self, captain_card]]
      @hand.delete(captain_card)
      return true
    end

    pl.merchant_pirates[merchant_card].each do |pair|
      if pair[1].instance_of?(Admiral)
        next
      end

      if pair[1].color == captain_card.color and pair[0] != self
        return false
      end
    end

    pl.merchant_pirates[merchant_card].each do |pair|
      if pair[0] == self and captain_card.color != pair[1].color
        return false
      end
    end

    pl.merchant_pirates[merchant_card].append([self, captain_card])
    @hand.delete(captain_card)
    return true
  end

  def play_admiral(admiral_card, merchant_card)
    if not @hand.include?(admiral_card)
      return false
    end

    if not merchant_card.instance_of?(MerchantShip)
        return false
    end

    if not @merchant_ships_at_sea.include?(merchant_card)
      return false
    end

    if not admiral_card.instance_of?(Admiral)
      return false
    end

    if not @merchant_pirates.key?(merchant_card)
      @merchant_pirates[merchant_card] = [[self, admiral_card]]
      @hand.delete(admiral_card)
      return true
    end

    @merchant_pirates[merchant_card].append([self, admiral_card])
    @hand.delete(admiral_card)
    return true
  end
end

class Deck
  @@instance = nil
    attr_writer :instance

  def initialize
    if @@instance.nil?
      @cards = []
      # Add Admiral
      @cards.push(Admiral.get_instance())

      # Add 5 twos, 6 threes, 5 fours, 5 fives, 2 sixes, 1 seven, and 1 eight merchant cards
      (1..5).each do
        @cards.push(MerchantShip.new(2))
        @cards.push(MerchantShip.new(4))
        @cards.push(MerchantShip.new(5))
        @cards.push(MerchantShip.new(3))
      end

      @cards.push(MerchantShip.new(3))
      @cards.push(MerchantShip.new(7))
      @cards.push(MerchantShip.new(8))
      @cards.push(MerchantShip.new(6))
      @cards.push(MerchantShip.new(6))

      # Add four captains
      @cards.push(Captain.new("blue"))
      @cards.push(Captain.new("green"))
      @cards.push(Captain.new("purple"))
      @cards.push(Captain.new("gold"))

      # Add pirate ship cards 2 ones, 4 twos, 4 threes, and 2 fours for each color
      ["blue", "green", "purple", "gold"].each do |i|
        @cards.push(PirateShip.new(i, 1))
        @cards.push(PirateShip.new(i, 1))

        @cards.push(PirateShip.new(i, 2))
        @cards.push(PirateShip.new(i, 2))
        @cards.push(PirateShip.new(i, 2))
        @cards.push(PirateShip.new(i, 2))

        @cards.push(PirateShip.new(i, 3))
        @cards.push(PirateShip.new(i, 3))
        @cards.push(PirateShip.new(i, 3))
        @cards.push(PirateShip.new(i, 3))

        @cards.push(PirateShip.new(i, 4))
        @cards.push(PirateShip.new(i, 4))
      end

    else
      raise RuntimeError
    end
  end

  attr_accessor :cards

  def self.get_instance
    if @@instance.nil?
      @@instance = Deck.new()
    end
    return @@instance
  end

  def self.instance=(v)
    @@instance = v
  end
end

class Game
  def initialize(deck)
    @deck = deck
    @players = []
    @current_player = nil
  end

  attr_reader :deck, :players, :current_player
  attr_writer :current_player

  def create_players(names, min_players, max_players)
    if names.length() < min_players or names.length() > max_players
      raise RuntimeError
    end

    names.each do |name|
      @players.push(Player.new(name))
    end

  end

  def random_player
    if @players.length() == 0
      return nil
    end

    ran = Random.rand(@players.length())
    return @players[ran]
  end

  def start
    if @players.length() == 0
      return nil
    end

    dealer_ind = -1
    @players.each_with_index do |player, ind|
      if player.dealer == true
        dealer_ind = ind
      end
    end

    return @players[dealer_ind - 1]
  end

  def next
    if @players.length() == 0
      return nil
    end

    current_player_ind = -1
    @players.each_with_index do |player, ind|
      if player == @current_player
        current_player_ind = ind
      end
    end

    next_ind = current_player_ind + 1
    # wrap around
    if current_player_ind == (@players.length() - 1)
      next_ind = 0
    end

    return @players[next_ind]
  end

  def choose_player(pos)
    if pos > @players.length() or pos <= 0
      return nil
    end
    return @players[pos - 1]

  end

  def capture_merchant_ships
    @players.each do |player|
      player.merchant_ships_at_sea.each do |ship|

        if not player.merchant_pirates.key?(ship)
          player.merchant_ships_captured.append(ship)
          player.merchant_ships_at_sea.delete(ship)
          player.merchant_pirates.delete(ship)
          next
        end

        # Admiral & Captain case
        last_card = player.merchant_pirates[ship][-1][1]
        if last_card.instance_of?(Admiral) or last_card.instance_of?(Captain)
          last_player = player.merchant_pirates[ship][-1][0]
          last_player.merchant_ships_captured.append(ship)
          player.merchant_ships_at_sea.delete(ship)
          player.merchant_pirates.delete(ship)
          next
        end

        # Check pirates
        player_strength = {}
        player.merchant_pirates[ship].each do |attack_plyr, attack_card|
          if attack_card.instance_of?(PirateShip)
            if not player_strength.include?(attack_plyr)
              player_strength[attack_plyr] = attack_card.attack_value
            elsif player_strength.include?(attack_plyr)
              player_strength[attack_plyr] = player_strength[attack_plyr] + attack_card.attack_value
            end
          end
        end

        highest_player = []
        max_score = player_strength.values().max()
        player_strength.each do |plyr, scr|
          if scr == max_score
            highest_player.append(plyr)
          end
        end

        if highest_player.length() != 1
          next
        end

        highest_player[0].merchant_ships_captured.append(ship)
        player.merchant_ships_at_sea.delete(ship)
        player.merchant_pirates.delete(ship)
      end
    end
  end

  def show_winner
    player_scores = {}
    @players.each do |player|
      score = 0
      player.merchant_ships_captured.each do |captured|
        score += captured.value
      end
      player_scores[player] = score
    end

    winners = []
    max_score = player_scores.values().max()
    if max_score == 0
      return winners
    end

    player_scores.each do |plyr, scr|
      if scr == max_score
        winners.append([plyr, scr])
      end
    end

    return winners
  end
end

File: HW4/loot.py
########## IMAN ALI ##########
########## 112204305 #############
########## imaali #############

import random

pirateColors = ["blue", "green", "purple", "gold"]
maxPlayersAllowed = 5
minPlayersAllowed = 2
playerNames = ["Joy", "Nan", "Sat"]


class MerchantShip:
    def __init__(self, value):
        self.value = value

    def get_value(self):
        return self.value


class PirateShip:
    def __init__(self, color, attack_value):
        self.color = color
        self.attack_value = attack_value

    def get_value(self):
        return self.attack_value

    def get_color(self):
        return self.color


class Captain:
    def __init__(self, color):
        self.color = color

    def get_color(self):
        return self.color


class Admiral:
    __instance__ = None

    def __init__(self):
        if Admiral.__instance__ is None:
            Admiral.__instance__ = self
        else:
            raise RuntimeError("You cannot create another Admiral")

    @staticmethod
    def get_instance():
        if not Admiral.__instance__:
            Admiral()
        return Admiral.__instance__


class Player:
    def __init__(self, name: str):
        self.name = name
        self.merchant_ships_at_sea = []
        self.merchant_ships_captured = []
        self.hand = []
        self.merchant_pirates = {}
        self.dealer = False

    # Assign player as dealer, shuffle and distribute 6 cards to each player
    def deal(self, game_state):
        self.dealer = True
        random.shuffle(game_state.deck.cards)
        card = 0
        for i in range(6):
            for player in game_state.players:
                current = game_state.deck.cards[card]
                player.hand.append(current)
                game_state.deck.cards.remove(current)
                card += 1

    # Stimulates player drawing card from the deck
    def draw_card(self, game_state):
        if len(game_state.deck.cards) > 0:
            card = game_state.deck.cards.pop()
            self.hand.append(card)
        else:
            raise Exception('Deck is empty! Cannot draw a card!')

    # Remove MerchantShip card from hand and float it
    def float_merchant(self, card):
        if not isinstance(card, MerchantShip):
            return False

        if card not in self.hand:
            return False

        self.hand.remove(card)
        self.merchant_ships_at_sea.append(card)
        return True

    # Attack a floating ship
    def play_pirate(self, pirate_card, merchant_card, p1):
        if pirate_card not in self.hand:
            return False

        if not isinstance(merchant_card, MerchantShip):
            return False

        if merchant_card not in p1.merchant_ships_at_sea:
            return False

        if not isinstance(pirate_card, PirateShip):
            return False

        # if nobody has attacked this ship before then attack with any color
        if merchant_card not in p1.merchant_pirates:
            p1.merchant_pirates[merchant_card] = [(self, pirate_card)]
            self.hand.remove(pirate_card)
            return True

        # Cant play pirate card after captain or admiral
        last_card = p1.merchant_pirates[merchant_card][- 1][1]
        if isinstance(last_card, Captain) or isinstance(last_card, Admiral):
            return False

        # Check if some other player has attacked with this color
        for pair in p1.merchant_pirates[merchant_card]:
            if pair[1].get_color() == pirate_card.get_color() and pair[0] != self:
                return False

        # Check if player itself has attacked previously and color is not same
        for pair in p1.merchant_pirates[merchant_card]:
            if pair[0] is self:
                if pirate_card.get_color() != pair[1].get_color():
                    return False

        # If player hasn't attacked before then attack
        p1.merchant_pirates[merchant_card].append((self, pirate_card))
        self.hand.remove(pirate_card)
        return True

    def play_captain(self, captain_card, merchant_card, p1):
        if captain_card not in self.hand:
            return False

        if not isinstance(merchant_card, MerchantShip):
            return False

        if merchant_card not in p1.merchant_ships_at_sea:
            return False

        if not isinstance(captain_card, Captain):
            return False

        # if nobody has attacked this ship before then attack with any color
        if merchant_card not in p1.merchant_pirates:
            p1.merchant_pirates[merchant_card] = [(self, captain_card)]
            self.hand.remove(captain_card)
            return True

        # Check if some other player has attacked with this color
        for pair in p1.merchant_pirates[merchant_card]:
            if isinstance(pair[1], Admiral):
                continue

            if pair[1].get_color() == captain_card.get_color() and pair[0] != self:
                return False

        # Check if player itself has attacked previously and color is same
        for pair in p1.merchant_pirates[merchant_card]:
            if pair[0] is self:
                if captain_card.get_color() != pair[1].get_color():
                    return False

        p1.merchant_pirates[merchant_card].append((self, captain_card))
        self.hand.remove(captain_card)
        return True

    def play_admiral(self, admiral_card, merchant_card):
        if admiral_card not in self.hand:
            return False

        if merchant_card not in self.merchant_ships_at_sea:
            return False

        if not isinstance(admiral_card, Admiral):
            return False

        if not isinstance(merchant_card, MerchantShip):
            return False

        # if nobody has attacked this ship before then attack
        if merchant_card not in self.merchant_pirates:
            self.merchant_pirates[merchant_card] = [(self, admiral_card)]
            self.hand.remove(admiral_card)
            return True

        self.merchant_pirates[merchant_card].append((self, admiral_card))
        self.hand.remove(admiral_card)
        return True


class Deck:
    __instance__ = None

    def __init__(self):
        if Deck.__instance__ is None:
            Deck.__instance__ = self
            self.cards = []

            # Add Admiral
            self.cards.append(Admiral.get_instance())  # 0

            # Add 5 twos, 6 threes, 5 fours, 5 fives, 2 sixes, 1 seven, and 1 eight merchant cards
            for i in range(5):
                self.cards.append(MerchantShip(2))  # 1,5,9,13,17
                self.cards.append(MerchantShip(4))  # 2,6,10,14,18
                self.cards.append(MerchantShip(5))  # 3,7,11,15,19
                self.cards.append(MerchantShip(3))  # 4,8,12,16,20

            self.cards.append(MerchantShip(3))  # 21
            self.cards.append(MerchantShip(7))  # 22
            self.cards.append(MerchantShip(8))  # 23
            self.cards.append(MerchantShip(6))  # 24
            self.cards.append(MerchantShip(6))  # 25

            # Add four captains
            self.cards.append(Captain("blue"))  # 26
            self.cards.append(Captain("green"))  # 27
            self.cards.append(Captain("purple"))  # 28
            self.cards.append(Captain("gold"))  # 29

            # Add pirate ship cards 2 ones, 4 twos, 4 threes, and 2 fours for each color
            for i in ["blue", "green", "purple", "gold"]:
                self.cards.append(PirateShip(i, 1))  # 30,42,54,66
                self.cards.append(PirateShip(i, 1))  # 31,43,55,67

                self.cards.append(PirateShip(i, 2))  # 32,44,56,68
                self.cards.append(PirateShip(i, 2))  # 33,45,57,69
                self.cards.append(PirateShip(i, 2))  # 34,46,58,70
                self.cards.append(PirateShip(i, 2))  # 35,47,59,71

                self.cards.append(PirateShip(i, 3))  # 36,48,60,72
                self.cards.append(PirateShip(i, 3))  # 37,49,61,73
                self.cards.append(PirateShip(i, 3))  # 38,50,62,74
                self.cards.append(PirateShip(i, 3))  # 39,51,63,75

                self.cards.append(PirateShip(i, 4))  # 40,52,64,76
                self.cards.append(PirateShip(i, 4))  # 41,53,65,77

        else:
            raise RuntimeError("You cannot create another Deck")

    @staticmethod
    def get_instance():
        if not Deck.__instance__:
            Deck()
        return Deck.__instance__


class Game:
    def __init__(self, deck):
        self.deck = deck
        self.players = []
        self.current_player = None

    # Creates a list of players instances with given names
    def create_players(self, names, min_players, max_players):
        if len(names) < min_players or len(names) > max_players:
            raise RuntimeError

        for name in names:
            self.players.append(Player(name))

    # Returns random player
    def random_player(self):
        if not self.players:
            return None

        rand = random.randint(0, len(self.players) - 1)
        return self.players[rand]

    # Returns player to the left of the dealer
    def start(self):
        dealer_ind = -1
        for ind, player in enumerate(self.players):
            if player.dealer:
                dealer_ind = ind

        return self.players[dealer_ind - 1]

    # Returns player to right of current player
    def next(self):
        current_player_ind = -1
        for ind, player in enumerate(self.players):
            if player is self.current_player:
                current_player_ind = ind

        next_ind = current_player_ind + 1
        # wrap around
        if current_player_ind == (len(self.players) - 1):
            next_ind = 0

        return self.players[next_ind]

    # Returns player at pos (1-indexed)
    def choose_player(self, pos):
        if pos > len(self.players) or pos <= 0:
            return None
        return self.players[pos - 1]

    # Assign merchant ships at sea to appropriate winner at end of round
    def capture_merchant_ships(self):
        # Check every merchant ship floated
        for player in self.players:
            for ship in player.merchant_ships_at_sea:

                # Never attacked
                if ship not in player.merchant_pirates:
                    player.merchant_ships_captured.append(ship)
                    player.merchant_ships_at_sea.remove(ship)
                    continue

                # Admiral & Captain case
                last_card = player.merchant_pirates[ship][-1][1]
                if isinstance(last_card, Admiral) or isinstance(last_card, Captain):
                    last_player = player.merchant_pirates[ship][-1][0]
                    last_player.merchant_ships_captured.append(ship)
                    player.merchant_ships_at_sea.remove(ship)
                    del player.merchant_pirates[ship]
                    continue

                # Check pirates
                player_strength = {}

                for (attack_plyr, attack_card) in player.merchant_pirates[ship]:
                    if isinstance(attack_card, PirateShip):
                        if attack_plyr not in player_strength:
                            player_strength[attack_plyr] = attack_card.get_value()

                        elif attack_plyr in player_strength:
                            player_strength[attack_plyr] = player_strength[attack_plyr] + attack_card.get_value()

                highest_player = []
                max_score = max(player_strength.values())
                for plyr, strength in player_strength.items():
                    if strength == max_score:
                        highest_player.append(plyr)

                # No winner. Tie!
                if len(highest_player) != 1:
                    continue

                highest_player[0].merchant_ships_captured.append(ship)
                player.merchant_ships_at_sea.remove(ship)
                del player.merchant_pirates[ship]

    # List of all winner with highest score
    def show_winner(self):
        player_scores = {}
        for player in self.players:
            score = 0
            for captured in player.merchant_ships_captured:
                score += captured.get_value()

            player_scores[player] = score

        winners = []
        max_score = max(player_scores.values())
        if max_score == 0:
            return winners

        for plyr, scr in player_scores.items():
            if scr == max_score:
                winners.append((plyr, scr))

        return winners
